runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,264 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+   }
+
 ///////////////////////////////////////////////////////////////////////////////////////
 // multianewArrayEvaluator:  multi-dimensional new array of objects
 ///////////////////////////////////////////////////////////////////////////////////////
 TR::Register *
 J9::Z::TreeEvaluator::multianewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    {
-   TR::ILOpCodes opCode = node->getOpCodeValue();
-   TR::Node::recreate(node, TR::acall);
-   TR::Register * targetRegister = directCallEvaluator(node, cg);
-   TR::Node::recreate(node, opCode);
-   return targetRegister;
-   }
+   static char *useDirectHelperCall = feGetEnv("TR_MultiANewArrayEvaluatorUseDirectCall");
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp->fe());
+   TR::Register *targetReg = cg->allocateRegister();
+
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild = node->getThirdChild();
+
+   if (useDirectHelperCall || !secondChild->getOpCode().isLoadConst() || secondChild->getInt() != 2)
+      {
+      traceMsg(cg->comp(), "multianewArrayEvaluator - More than 2 dimensions, so we can't handle this scenario in JITT'ed code\n");
+      return performHelperCall(node, NULL, TR::acall, false, cg);
+      }
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *nonZeroFirstDimLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *oolFailLabel = generateLabelSymbol(cg);
+
+   // oolJumpLabel is a common point that all branches will jump to. From this label, we branch to OOL code.
+   // We do this instead of jumping directly to OOL code from mainline because the RA can only handle the case where there's
+   // a single jump point to OOL code.
+   TR::LabelSymbol *oolJumpLabel = generateLabelSymbol(cg);
+
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+
+   TR::Register *dimsPtrReg = cg->evaluate(firstChild);
+   TR::Register *dimReg = cg->evaluate(secondChild);
+   TR::Register *classReg = cg->evaluate(thirdChild);
+
+   // In the mainline, first load the first and second dimensions' lengths into registers.
+   TR::Register *firstDimLenReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::L, node, firstDimLenReg, generateS390MemoryReference(dimsPtrReg, 4, cg));
+
+   TR::Register *secondDimLenReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::L, node, secondDimLenReg, generateS390MemoryReference(dimsPtrReg, 0, cg));
+
+   // Check to see if second dimension is indeed 0. If yes, then proceed to handle the case here. Otherwise jump to OOL code.
+   generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CL, node, secondDimLenReg, 0, TR::InstOpCode::COND_BNE, oolJumpLabel, false);
```

Comment:
Does this mean we only allow inline allocations if the signature is `[x][0]`? That doesn't really make sense to me.

Reply:
Would you be able to elaborate on why it doesn't make sense?

The purpose of doing this work on x86 originally was to only accelerate cases where the second dimension is 0: https://github.com/eclipse/openj9/pull/2408#issue-202022369. The goal was always to only accelerate these cases as performance investigations showed that some benchmarks were generating 2 dimensional arrays with the second dimension being 0.

So I'm doing a similar thing on Z.

Reply:
Just read over the x86 comments you referenced. It seems odd that we restricted to only zero length second dimensions. I don't think it would have been much extra work to support the general case. But alas it seems this change was driven by what we saw in the profiles of a benchmark. I suppose there was no need to accelerate non-zero second dimensions, at least at the time.

I was under the assumption that we are doing the general case, which is why it didn't make sense that we restricted only to `[x][0]` here. My assumption was wrong as you point out. Thanks for linking that comment. This part looks fine.